crates/viewer/re_viewer_context/src/cache/video_stream_cache.rs

```diff
@@ -0,0 +1,555 @@
+            chunk,
+            codec,
+            &mut video_sample_buffers,
+            &mut samples,
+            &mut gops,
+        )?;
+    }
+
+    Ok((
+        re_video::VideoDataDescription {
+            codec,
+            stsd: None,                               // Only on mp4s have this.
+            coded_dimensions: None,                   // Unknown so far.
+            timescale: re_video::Timescale::IDENTITY, // We use raw rerun timestamps, so we don't have to scale time.
+            duration: None, // Streams have to be assumed to be open ended, so we don't have a duration.
+            gops,
+            samples,
+            samples_statistics: re_video::SamplesStatistics::NO_BFRAMES, // TODO(#10090): No b-frames for now.
+            mp4_tracks: std::iter::once((0, Some(re_video::TrackKind::Video))).collect(),
+        },
+        video_sample_buffers,
+    ))
+}
+
+/// Reads all samples from a chunk and appends them to a list.
+///
+/// Rejects out of order samples - new samples must have a higher timestamp than the previous ones.
+fn read_samples_from_chunk(
+    timeline: TimelineName,
+    frame_chunk_descr: &re_types::ComponentDescriptor,
+    chunk: &re_chunk::Chunk,
+    codec: re_video::VideoCodec,
+    chunk_buffers: &mut StableIndexDeque<SampleBuffer>,
+    samples: &mut StableIndexDeque<re_video::SampleMetadata>,
+    gops: &mut StableIndexDeque<re_video::demux::GroupOfPictures>,
+) -> Result<(), VideoStreamProcessingError> {
+    re_tracing::profile_function!();
+
+    let Some(raw_array) = chunk.raw_component_array(frame_chunk_descr) else {
+        // This chunk doesn't have any video chunks.
+        return Ok(());
+    };
+
+    let inner_list_array = raw_array
+        .downcast_array_ref::<arrow::array::ListArray>()
+        .ok_or(VideoStreamProcessingError::InvalidVideoSampleType(
+            raw_array.data_type().clone(),
+        ))?;
+    let values = inner_list_array
+        .values()
+        .downcast_array_ref::<arrow::array::PrimitiveArray<arrow::array::types::UInt8Type>>()
+        .ok_or(VideoStreamProcessingError::InvalidVideoSampleType(
+            raw_array.data_type().clone(),
+        ))?;
+    let values = values.values().inner();
+
+    let offsets = inner_list_array.offsets();
+    let lengths = re_arrow_util::offsets_lengths(inner_list_array.offsets()).collect::<Vec<_>>();
+
+    let buffer_index = chunk_buffers.next_index();
+    let sample_base_idx = samples.next_index();
+
+    let mut previous_max_presentation_timestamp = samples
+        .back()
+        .map_or(re_video::Time::MIN, |s| s.presentation_timestamp);
+
+    // Extract sample metadata.
+    samples.extend(
+        chunk
+            .iter_component_offsets(frame_chunk_descr)
+            .zip(chunk.iter_component_indices(&timeline, frame_chunk_descr))
+            .filter_map(move |((idx, len), (time, _row_id))| {
+                debug_assert_eq!(len, 1, "Expected only a single video sample per timestep");
```

Comment:
It would be much better to error here if `len != 1`. Right now the code may crash if this invariant isn't held, even in release builds.